controller/src/main/java/io/pravega/controller/store/stream/PravegaTablesStream.java

```diff
@@ -278,15 +277,32 @@ private String getWritersTableName(String id) {
     }
 
     @Override
-    public CompletableFuture<Void> createSubscriber(String newSubscriber) {
+    public CompletableFuture<Void> createSubscriber(String newSubscriber, long operationGeneration) {
         final StreamSubscriber newSubscriberRecord = new StreamSubscriber(newSubscriber, ImmutableMap.of(), System.currentTimeMillis());
         return getMetadataTable()
                 .thenCompose(metadataTable -> getSubscriberSetRecord(true)
-                        .thenCompose(subscriberSetRecord -> storeHelper.updateEntry(metadataTable, SUBSCRIBER_SET_KEY,
-                                SubscriberSet.add(subscriberSetRecord.getObject(), newSubscriber).toBytes(), subscriberSetRecord.getVersion())
-                                .thenCompose(v -> storeHelper.addNewEntryIfAbsent(metadataTable, getKeyForSubscriber(newSubscriber), newSubscriberRecord.toBytes()))
-                                .thenAccept(v -> storeHelper.invalidateCache(metadataTable, SUBSCRIBER_SET_KEY))
-                                .thenAccept(v -> storeHelper.invalidateCache(metadataTable, getKeyForSubscriber(newSubscriber)))));
+                   .thenCompose(subscriberSetRecord -> {
+                       if (subscriberSetRecord.getObject().getSubscribers().containsKey(newSubscriber)) {
+                           // update Subscriber generation
+                           Long generation = subscriberSetRecord.getObject().getSubscribers().get(newSubscriber);
+                           if (generation.longValue() < operationGeneration) {
+                               storeHelper.updateEntry(metadataTable, SUBSCRIBER_SET_KEY,
```

Comment:
this is a hanging future. this should be returned from this thenCompose call

Reply:
fixed